File: iosApp/iosApp/ui/components/state/ManagementResourceState.swift
//
//  BaseView.swift
//  iosApp
//
//  Created by Daniel Ávila on 25/12/21.
//  Copyright © 2021 orgName. All rights reserved.
//

import SwiftUI
import shared

struct ManagementResourceState<T : AnyObject, SuccessView : View>: View {
    private let resourceState: BasicUiState
    private let successView: (T?) -> SuccessView
    private let onTryAgain: () -> Void
    private let msgTryAgain: String
    private let onCheckAgain: () -> Void
    private let msgCheckAgain: String

    init(
        resourceState: BasicUiState,
        successView:  @escaping (T?) -> SuccessView,
        onTryAgain: @escaping () -> Void,
        msgTryAgain: String = "An error has ocurred",
        onCheckAgain: @escaping () -> Void,
        msgCheckAgain: String = "No data to show"
    ) {
        self.resourceState = resourceState
        self.successView = successView
        self.onTryAgain = onTryAgain
        self.msgTryAgain = msgTryAgain
        self.onCheckAgain = onCheckAgain
        self.msgCheckAgain = msgCheckAgain
    }
    
    var body: some View {
        switch self.resourceState {
            case is BasicUiStateEmpty:
                EmptyView(msg: self.msgCheckAgain, onClick: self.onCheckAgain)
            case is BasicUiStateError:
                ErrorView(msg: self.msgTryAgain, onClick: self.onTryAgain)
            case is BasicUiStateLoading:
                LoadingView()
            case let success as BasicUiStateSuccess<T>:
                successView(success.data)
            default:
                Text("You shouldn't be seeing this message")
        }
    }
}

File: composeApp/src/commonMain/kotlin/daniel/avila/rnm/kmm/presentation/ui/common/state/ManagementResourceState.kt
package daniel.avila.rnm.kmm.presentation.ui.common.state

import androidx.compose.foundation.layout.Box
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import daniel.avila.rnm.kmm.presentation.mvi.BasicUiState

@Composable
fun <T> ManagementResourceState(
    modifier: Modifier = Modifier,
    resourceState: BasicUiState<T>,
    successView: @Composable (data: T) -> Unit,
    loadingView: @Composable () -> Unit = { Loading() },
    onTryAgain: () -> Unit,
    msgTryAgain: String = "No data to show",
    onCheckAgain: () -> Unit,
    msgCheckAgain: String = "An error has ocurred"
) {
    Box(
        modifier = modifier,
        contentAlignment = Alignment.Center,
    ) {
        when (resourceState) {
            is BasicUiState.Empty -> Empty(modifier = modifier, onCheckAgain = onCheckAgain, msg = msgCheckAgain)
            is BasicUiState.Error -> Error(modifier = modifier, onTryAgain = onTryAgain, msg = msgTryAgain)
            BasicUiState.Loading -> loadingView()
            is BasicUiState.Success -> successView(resourceState.data)
            BasicUiState.Idle -> Unit
        }
    }
}